Adding support for uuid in url.

artursouza opened the pull request:
Closes #425

artursouza committed:
Adding support for uuid in url in stdclient.

artursouza commented:
This feature seems to only be possible in the standard client mode. For fast client, it would slow it down a lot since a new request byte array would need to be recomputed per call.

ldemailly commented:
Thanks a lot for contributing

See a question in https://github.com/fortio/fortio/issues/425#issuecomment-722520027

You could support the fast client because the uuid is actually fixed length so you could just poke the bytes

Did you run some self benchmark? what's the slowdown when using uuid? could it be a simple sequence number instead?

Also... tests  :-)

artursouza commented:
I will see if there is a way to generate this in the fast client too. There might be a dummy zero UUID in the beginning to mark the byte positions before it starts generating unique ones for each requests.

I like the idea to have ids too. Maybe users can decide which one they need and know that uuid will have a performance hit of some sort.

I did not run a benchmark to compare. I did run it for our scenario and I had enough qps for our load test.

artursouza committed:
Add uuid to fastClient.

ldemailly commented:
day work has taken over a bit so I am behind looking at this, thanks though, but can you add tests?

artursouza committed:
Unit test for uuid.

artursouza commented:
I've added unit tests. I am pasting some benchmarks below. Not sure what is the success criteria here.

artursouza commented:
Here is a summary of my test:

## Hardware

* MacBook Pro (16-inch, 2019)
* 2.6 GHz 6-Core Intel Core i7
* 16 GB 2667 MHz DDR4
* macOS Catalina 10.15.7

## Test configuration
Requested QPS: 10k
Duration: 1 minute
Threads/Simultaneous connections: 1

| HEAD  | Client | URL | Actual QPS
| - | - | - | - |
| 7c6866 | Fast | http://localhost:8080/echo/{uuid}?delay=250us:30,5ms:5\u0026status=503:0.5,429:1.5 | 2254
| 7c6866  | Standard  | http://localhost:8080/echo/{uuid}?delay=250us:30,5ms:5\u0026status=503:0.5,429:1.5 | 2208
| 7c6866 | Fast | http://localhost:8080/echo?delay=250us:30,5ms:5\u0026status=503:0.5,429:1.5 | 2287
| 7c6866  | Standard  | http://localhost:8080/echo?delay=250us:30,5ms:5\u0026status=503:0.5,429:1.5 | 2044
| master | Fast | http://localhost:8080/echo?delay=250us:30,5ms:5\u0026status=503:0.5,429:1.5 | 2258
| master  | Standard  | http://localhost:8080/echo?delay=250us:30,5ms:5\u0026status=503:0.5,429:1.5 | 2199

ldemailly commented:
Thanks for the tests!!

to test raw/max perf the best way is to use qps -1 (max qps) and c = 2xcores and no delay (ie no arguments except your uuid in the echo url)

ie something like

http://localhost:8080/fortio/?qps=-1&t=10s&c=10&url=http://localhost:8080/echo&save=on&load=Start
(removed the params that are optional/default)

but I'll checkout your branch and check, no worries

ldemailly commented:
ideally this would work in payload too and not just query args

artursouza commented:
> ideally this would work in payload too and not just query args

This can work in payload too. I can create a separate issue for that.

ldemailly commented on fhttp/http_test.go:
```diff
@@ -841,6 +842,58 @@ func TestPayloadSizeLarge(t *testing.T) {
+	}
+}
+
+func TestBadUUIDFastClient(t *testing.T) {
```
very good to have a negative test

ldemailly commented on fhttp/http_test.go:
```diff
@@ -1190,6 +1243,23 @@ func TestInsecureRequestWithResolve(t *testing.T) {
 }
 
+// ValidateUUIDPath is an http server handler validating /{uuid}.
+func ValidateUUIDPath(w http.ResponseWriter, r *http.Request) {
```
👍

ldemailly commented on fhttp/http_test.go:
```diff
@@ -841,6 +842,58 @@ func TestPayloadSizeLarge(t *testing.T) {
+	url := fmt.Sprintf("http://localhost:%d/{uuid}", a.Port)
+	o := HTTPOptions{URL: url, DisableFastClient: false}
+	client, _ := NewClient(&o)
+	code, data, header := client.Fetch()
```
one missing test though is that if you make a 2nd request you get a different UUID in the context of the httprunner (you could set num request to 10, connections to 1 or 2 and check you get all different results)

artursouza replied:
Done.

ldemailly reviewed (commented):
not done with review, need to checkout and see

ldemailly commented:
it works with fast client:

09:48:08 I http_server.go:413> Echo: GET /echo?uuid={aced2e42-4b40-42c4-8e97-2449bb17c0a6} HTTP/1.1 [::1]:64067 ()
09:48:08 I http_server.go:413> Echo: GET /echo?uuid={1b33f4ba-5bd1-452c-b891-1ac66aea6824} HTTP/1.1 [::1]:64069 ()
...

but it doesn't work with std client:

09:48:46 I http_server.go:413> Echo: GET /echo?uuid={{uuid}} HTTP/1.1 [::1]:64083 ()
09:48:46 I http_server.go:413> Echo: GET /echo?uuid={{uuid}} HTTP/1.1 [::1]:64090 ()


Try with 
fortio server -loglevel verbose

and visit
http://localhost:8080/fortio/?url=http%3A%2F%2Flocalhost%3A8080%2Fecho%3Fuuid%3D%7B%7Buuid%7D%7D&n=3&c=1&runner=http&stdclient=on&load=Start

That's why a test using runner, but also a manual test like I just did is useful

ldemailly commented on fhttp/http_client.go:
```diff
@@ -46,6 +47,10 @@ type Fetcher interface {
 }
 
+const (
+	urlUUIDToken = "{uuid}"
```
change to UUIDToken so that's one less change for when we use it in payload too

ldemailly commented on fhttp/http_client.go:
```diff
@@ -409,8 +419,10 @@ func NewStdClient(o *HTTPOptions) (*Client, error) {
-		url: o.URL,
-		req: req,
+		url:          o.URL,
+		path:         req.URL.Path,
```
I suspect it's not working because you change the path and not the full url

see comment about test

artursouza replied:
Correct. I will fix that. I did not reproduce this problem because our use case only needed path to be replaced.

ldemailly reviewed (dismissed):
it'd be better if it worked :)

artursouza commented:
OK. I will work on adding more test cases and fix the issue this weekend. Thanks for testing this change.

ldemailly commented:
Btw I also noticed on a long run a long tail in latency only with uuid which probably comes from using crypto / secure rand for it (and needing more entropy regularly I suspect) we need to switch to math/rand for performance.

artursouza commented:
> Btw I also noticed on a long run a long tail in latency only with uuid which probably comes from using crypto / secure rand for it (and needing more entropy regularly I suspect) we need to switch to math/rand for performance.

Thanks for the insight. I will look at that as well.

ldemailly commented:
> > Btw I also noticed on a long run a long tail in latency only with uuid which probably comes from using crypto / secure rand for it (and needing more entropy regularly I suspect) we need to switch to math/rand for performance.
> 
> Thanks for the insight. I will look that as well.

most welcome. we have to make a small reader wrapper - I got as far as this but had to do other things
https://github.com/golang/go/issues/16776#issuecomment-737308147

artursouza committed:
Supports query. Supports multiple uuids. Uses math rand.

artursouza commented:
@ldemailly  I've addressed all the comments. I've identified one more issue. I am fixing it and modifying one of the tests.

artursouza commented:
@ldemailly It should be good to perf test. Can you run the same perf tests again?

artursouza commented:
BTW, I've also added support for multiple UUIDs, so URLs can be like:
```
19:24:20 I http_server.go:413> Echo: GET /echo/983ad889-9c1a-4081-9294-5f1c1174b020/?uuid=968c8070-e3d4-4e88-a8bd-a4d053748d4e 
```

ldemailly commented:
it's been a very busy week at work but I hope to get to your MR tomorrow or this weekend

ldemailly commented:
ps: if possible in general it's best to avoid force pushing

ldemailly commented:
I had a bit of time and was able to test performance and it's much better with the math.random (despite the somewhat expensive mutex)

will look at code next, thanks for your patience

artursouza commented:
Thanks. I force push only when I made a silly mistake on the latest commit and it has not passed "too much" time :)

artursouza committed:
Fix standard client with many uuids in the path.

artursouza commented:
I've rebased it with master and fixed one bug where the standard client did not handle multiple uuids in the path - unit test was modified to cover that too.

ldemailly commented:
numbers are good now (~60k qps with fast client on my mac with hardcoded or generated uuids)

just need to look at the code and also see how easy it is to add the substitution to the "payload" (should be easy if you already have multiple substitution support)

artursouza commented:
NP. I can take a look at payload substitution too. I will update the PR with this over the weekend.

ldemailly commented:
> NP. I can take a look at payload substitution too. I will update the PR with this over the weekend.

thanks ! that would be great!

ldemailly commented:
and for the fast client the buffer is a blob including payload so the byte offets you came up with should work just fine there as well

artursouza committed:
Adds support for {uuid} in payload.

artursouza commented:
I've added support for "{uuid}" in the payload too. thanks.

ldemailly reviewed (approved):
for the fast client I think the substitution/buffer replacement could be done in 1 pass for all {{uuid}} as there is no difference between what is in the URL and was is elsewhere in the payload; actually even headers could have UUID too (like traceidd etc) so a wholesale search and replace would work even better?

(this being said this has been open for so long (my apologies) that maybe we can refine this later)

artursouza commented:
@ldemailly Thanks. I thought about doing the replacement in one pass, the only problem is that it would contain the same generated uuid instead of one per replacement.

Can this PR be merged? I don't have merge permission.

ldemailly commented:
Thanks a lot for your work on this, merging it in and it will be in fortio 1.12
